{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb5400b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from datasets import Dataset, load_dataset\n",
    "from pandas import DataFrame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e1bcbc6",
   "metadata": {},
   "source": [
    "# Load Dataset From HuggingFace Hub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ff03d93",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = load_dataset(\"GustavoHCruz/DNA_Coding_Regions\", split=\"train\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aef96fb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(dataset, Dataset)\n",
    "\n",
    "df = dataset.to_pandas()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a13d3ca",
   "metadata": {},
   "source": [
    "## Exons and Introns Binary Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ed55da1",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(df, DataFrame)\n",
    "\n",
    "final_df = []\n",
    "for _, row in df.iterrows():\n",
    "\tdna = row[\"sequence\"]\n",
    "\torg = row[\"organism\"]\n",
    "\n",
    "\tfor exon in row[\"introns\"]:\n",
    "\t\tseq = exon.get(\"sequence\", \"\")\n",
    "\t\tgene = exon.get(\"gene\", \"\")\n",
    "\t\tbefore = exon.get(\"before\", \"\")\n",
    "\t\tafter = exon.get(\"after\", \"\")\n",
    "\n",
    "\t\tif len(seq) >= 256:\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tfinal_df.append({\n",
    "\t\t\t\"sequence\": seq,\n",
    "\t\t\t\"target\": \"INTRON\",\n",
    "\t\t\t\"organism\": org,\n",
    "\t\t\t\"gene\": gene,\n",
    "\t\t\t\"before\": before,\n",
    "\t\t\t\"after\": after\n",
    "\t\t})\n",
    "\n",
    "\tfor exon in row[\"exons\"]:\n",
    "\t\tseq = exon.get(\"sequence\", \"\")\n",
    "\t\tgene = exon.get(\"gene\", \"\")\n",
    "\t\tbefore = exon.get(\"before\", \"\")\n",
    "\t\tafter = exon.get(\"after\", \"\")\n",
    "\n",
    "\t\tif len(seq) >= 256:\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tfinal_df.append({\n",
    "\t\t\t\"sequence\": seq,\n",
    "\t\t\t\"target\": \"EXON\",\n",
    "\t\t\t\"organism\": org,\n",
    "\t\t\t\"gene\": gene,\n",
    "\t\t\t\"before\": before,\n",
    "\t\t\t\"after\": after\n",
    "\t\t})\n",
    "\n",
    "final_df = pd.DataFrame(final_df)\n",
    "\n",
    "final_df.to_csv(\"./storage/data/processed/exin-256.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5d2f20",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(df, DataFrame)\n",
    "\n",
    "final_df = []\n",
    "for _, row in df.iterrows():\n",
    "\tdna = row[\"sequence\"]\n",
    "\torg = row[\"organism\"]\n",
    "\n",
    "\tfor exon in row[\"introns\"]:\n",
    "\t\tseq = exon.get(\"sequence\", \"\")\n",
    "\t\tgene = exon.get(\"gene\", \"\")\n",
    "\t\tbefore = exon.get(\"before\", \"\")\n",
    "\t\tafter = exon.get(\"after\", \"\")\n",
    "\n",
    "\t\tif len(seq) >= 512:\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tfinal_df.append({\n",
    "\t\t\t\"sequence\": seq,\n",
    "\t\t\t\"target\": \"INTRON\",\n",
    "\t\t\t\"organism\": org,\n",
    "\t\t\t\"gene\": gene,\n",
    "\t\t\t\"before\": before,\n",
    "\t\t\t\"after\": after\n",
    "\t\t})\n",
    "\n",
    "\tfor exon in row[\"exons\"]:\n",
    "\t\tseq = exon.get(\"sequence\", \"\")\n",
    "\t\tgene = exon.get(\"gene\", \"\")\n",
    "\t\tbefore = exon.get(\"before\", \"\")\n",
    "\t\tafter = exon.get(\"after\", \"\")\n",
    "\n",
    "\t\tif len(seq) >= 512:\n",
    "\t\t\tcontinue\n",
    "\n",
    "\t\tfinal_df.append({\n",
    "\t\t\t\"sequence\": seq,\n",
    "\t\t\t\"target\": \"EXON\",\n",
    "\t\t\t\"organism\": org,\n",
    "\t\t\t\"gene\": gene,\n",
    "\t\t\t\"before\": before,\n",
    "\t\t\t\"after\": after\n",
    "\t\t})\n",
    "\n",
    "final_df = pd.DataFrame(final_df)\n",
    "\n",
    "final_df.to_csv(\"./storage/data/processed/exin-512.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59004464",
   "metadata": {},
   "source": [
    "## Nucleotide Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80231720",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_target(\n",
    "\tseq: str,\n",
    "\tintrons: list,\n",
    "\texons: list\n",
    ") -> str:\n",
    "\ttarget = np.array([\"U\"] * len(seq))\n",
    "\n",
    "\tfor exon in exons:\n",
    "\t\tstart = exon.get(\"start\")\n",
    "\t\tend = exon.get(\"end\")\n",
    "\t\ttarget[start:end] = \"E\"\n",
    "\n",
    "\tfor intron in introns:\n",
    "\t\tstart = intron.get(\"start\")\n",
    "\t\tend = intron.get(\"end\")\n",
    "\t\ttarget[start:end] = \"I\"\n",
    "\n",
    "\treturn \"\".join(target)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bb51b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(df, DataFrame)\n",
    "df = df[df[\"sequence\"].str.len() < 500]\n",
    "\n",
    "results = []\n",
    "for _, row in df.iterrows():\n",
    "\tseq = row[\"sequence\"]\n",
    "\torg = row[\"organism\"]\n",
    "\tintrons = row[\"introns\"]\n",
    "\texons = row[\"exons\"]\n",
    "\n",
    "\tif len(introns) == 0 and len(exons) == 0:\n",
    "\t\tcontinue\n",
    "\n",
    "\ttarget = build_target(seq, introns, exons)\n",
    "\n",
    "\tresults.append({\n",
    "\t\t\"sequence\": seq,\n",
    "\t\t\"target\": target,\n",
    "\t\t\"organism\": org\n",
    "\t})\n",
    "\n",
    "result_df = pd.DataFrame(results)\n",
    "result_df.to_csv(\"./storage/data/processed/nucl-500.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c927afe3",
   "metadata": {},
   "source": [
    "## DNA Translator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3ab80ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "def join_proteins(protein_list) -> str:\n",
    "\tseqs = []\n",
    "\tfor p in protein_list:\n",
    "\t\tseq = p.get(\"sequence\", \"\")\n",
    "\t\tif isinstance(seq, str):\n",
    "\t\t\tseqs.append(seq)\n",
    "\treturn \"\".join(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "169706ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(df, DataFrame)\n",
    "\n",
    "df[\"target\"] = df[\"proteins\"].apply(join_proteins).astype(str)\n",
    "mask = (df[\"sequence\"].str.len() + df[\"target\"].str.len()) < 500\n",
    "filtered_df = df[mask]\n",
    "\n",
    "final_df = filtered_df[[\"sequence\", \"target\", \"organism\"]]\n",
    "\n",
    "final_df.to_csv(\"./storage/data/processed/tran-500.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1ad9776",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(df, DataFrame)\n",
    "\n",
    "df[\"target\"] = df[\"proteins\"].apply(join_proteins).astype(str)\n",
    "mask = (df[\"sequence\"].str.len() + df[\"target\"].str.len()) < 1000\n",
    "filtered_df = df[mask]\n",
    "\n",
    "final_df = filtered_df[[\"sequence\", \"target\", \"organism\"]]\n",
    "\n",
    "final_df.to_csv(\"./storage/data/processed/tran-1000.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bb8edf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(df, DataFrame)\n",
    "\n",
    "df[\"target\"] = df[\"proteins\"].apply(join_proteins).astype(str)\n",
    "mask = (df[\"sequence\"].str.len() + df[\"target\"].str.len()) < 8000\n",
    "filtered_df = df[mask]\n",
    "\n",
    "final_df = filtered_df[[\"sequence\", \"target\", \"organism\"]]\n",
    "\n",
    "final_df.to_csv(\"./storage/data/processed/tran-8000.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e0204a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert isinstance(df, DataFrame)\n",
    "\n",
    "df[\"target\"] = df[\"proteins\"].apply(join_proteins).astype(str)\n",
    "mask = (df[\"sequence\"].str.len() + df[\"target\"].str.len()) < 16000\n",
    "filtered_df = df[mask]\n",
    "\n",
    "final_df = filtered_df[[\"sequence\", \"target\", \"organism\"]]\n",
    "\n",
    "final_df.to_csv(\"./storage/data/processed/tran-16000.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "llm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
