{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e11a9d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from itertools import islice\n",
    "from typing import Iterator\n",
    "\n",
    "from data_processor.raw_extractor import extract_data\n",
    "from datasets import Dataset\n",
    "from schemas.tables_data import DNASequence\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7684c342",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations_file_path = \"./storage/datasets/genbank/gb_curated.gb\"\n",
    "HF_DATASET_NAME = \"gu-dudi/DNA_coding_regions\"\n",
    "OUTPUT_JSONL = \"./storage/data/base/dna_coding_regions.jsonl\"\n",
    "CHUNK_SIZE = 2000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e5d8336",
   "metadata": {},
   "outputs": [],
   "source": [
    "generator = extract_data(\n",
    "\tannotations_file_path=annotations_file_path\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd9c8937",
   "metadata": {},
   "outputs": [],
   "source": [
    "def chunk_generator(\n",
    "\tgenerator: Iterator[DNASequence],\n",
    "\tchunk_size: int = 2000\n",
    ") -> Iterator[list[DNASequence]]:\n",
    "\titerator = iter(generator)\n",
    "\twhile batch := list(islice(iterator, chunk_size)):\n",
    "\t\tyield batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11a8558a",
   "metadata": {},
   "outputs": [],
   "source": [
    "seen_accession = set()\n",
    "seen_seq_org = set()\n",
    "\n",
    "skipped_accession = 0\n",
    "skipped_seq_org = 0\n",
    "skipped_invalid = 0\n",
    "\n",
    "with open(OUTPUT_JSONL, \"w\", encoding=\"utf-8\") as f_out:\n",
    "\tfor chunk in tqdm(chunk_generator(generator, chunk_size=CHUNK_SIZE), desc=\"Processing DNA\"):\n",
    "\t\tfor dna in chunk:\n",
    "\t\t\tacc = dna[\"accession\"].strip()\n",
    "\t\t\torg = dna[\"organism\"].strip()\n",
    "\t\t\tseq = dna[\"sequence\"].strip()\n",
    "\t\t\t\n",
    "\t\t\tkey_accession = acc\n",
    "\t\t\tkey_seq_org = (seq, org)\n",
    "\n",
    "\n",
    "\t\t\tif key_accession in seen_accession:\n",
    "\t\t\t\tskipped_accession += 1\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\tif key_seq_org in seen_seq_org:\n",
    "\t\t\t\tskipped_seq_org += 1\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\tseen_accession.add(key_accession)\n",
    "\t\t\tseen_seq_org.add(key_seq_org)\n",
    "\t\t\n",
    "\t\t\texin_list = dna.get(\"exin\", [])\n",
    "\t\t\tprotein_list = dna.get(\"cds\", [])\n",
    "\n",
    "\t\t\tif not exin_list and not protein_list:\n",
    "\t\t\t\tskipped_invalid += 1\n",
    "\t\t\t\tcontinue\n",
    "\n",
    "\t\t\trecord = {\n",
    "\t\t\t\t\"accession\": acc,\n",
    "\t\t\t\t\t\"organism\": org,\n",
    "\t\t\t\t\t\"sequence\": seq,\n",
    "\t\t\t\t\t\"introns\": [],\n",
    "\t\t\t\t\t\"exons\": [],\n",
    "\t\t\t\t\t\"proteins\": []\n",
    "\t\t\t}\n",
    "\t\t\t\n",
    "\t\t\tfor exin in exin_list:\n",
    "\t\t\t\titem = {\n",
    "\t\t\t\t\t\"sequence\": exin[\"sequence\"].strip(),\n",
    "\t\t\t\t\t\"start\": exin[\"start\"],\n",
    "\t\t\t\t\t\"end\": exin[\"end\"],\n",
    "\t\t\t\t\t\"gene\": exin.get(\"gene\", \"\"),\n",
    "\t\t\t\t\t\"before\": exin.get(\"before\"),\n",
    "\t\t\t\t\t\"after\": exin.get(\"after\")\n",
    "\t\t\t\t}\n",
    "\t\t\t\tif exin[\"type\"] == \"EXON\":\n",
    "\t\t\t\t\trecord[\"exons\"].append(item)\n",
    "\t\t\t\telif exin[\"type\"] == \"INTRON\":\n",
    "\t\t\t\t\trecord[\"introns\"].append(item)\n",
    "\t\t\t\t\n",
    "\t\t\tfor protein in protein_list:\n",
    "\t\t\t\trecord[\"proteins\"].append({\n",
    "\t\t\t\t\t\"sequence\": protein[\"sequence\"].strip(),\n",
    "\t\t\t\t\t\"start\": protein[\"start\"],\n",
    "\t\t\t\t\t\"end\": protein[\"end\"],\n",
    "\t\t\t\t\t\"gene\": protein.get(\"gene\", \"\")\n",
    "\t\t\t\t})\n",
    "\t\t\t\n",
    "\t\t\tf_out.write(json.dumps(record, ensure_ascii=False) + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2920b1fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Skipped {skipped_accession} duplicate accessions.\")\n",
    "print(f\"Skipped {skipped_seq_org} duplicate sequence+organism entries.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4d8f070",
   "metadata": {},
   "outputs": [],
   "source": [
    "def jsonl_to_dataset(\n",
    "\tpath,\n",
    "\tsample_limit = None\n",
    ") -> Dataset:\n",
    "\trecords = []\n",
    "\twith open(path, \"r\", encoding=\"utf-8\") as f:\n",
    "\t\tfor i, line in enumerate(f):\n",
    "\t\t\tif sample_limit and i >= sample_limit:\n",
    "\t\t\t\tbreak\n",
    "\t\t\trecords.append(json.loads(line))\n",
    "\t\n",
    "\treturn Dataset.from_list(records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dd53937",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = jsonl_to_dataset(OUTPUT_JSONL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cde9ea9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Dataset Len: {len(dataset)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2fb76947",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset.push_to_hub(HF_DATASET_NAME)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "llm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
